OpenZFS: 'vdev initialize' feature

behlendorf opened the pull request:
### Motivation and Context

Updated version of `vdev initialize`.  This PR builds on the work done by @grwilson and @dweeezil in
#7708, #7954, and #7955 to wrap up to remaining issues preventing this feature from being integrated.  The following two commits have been added:

2849493 Add 'zpool status -i' option
41158e8 Speed up vdev initialize stop/cancel

Commit 2849493 is entirely cosmetic but addresses a usability concern for administrators who manage their pools primarily via the CLI.  The `zpool status` command was updated to only output the 'vdev initialization' status when the `-i` option is provided.

Commit 41158e8 resolves the more substantial issues.  When initializing a pool with multiple leaf vdevs stopping/cancelling the operation could result in the spa_namespace_lock being held for many minutes. 
During which time no other CLI commands could be run making the pool appear to be hung.  This was resolved by restructuring the stop/cancel code.

### Description

This feature introduces a way to 'initialize' the disks at install or in the background to make sure we don't incur this first read penalty.

When an entire LUN is added to ZFS, we make all space available immediately, and allow ZFS to find unallocated space and zero it out. This works with concurrent writes to arbitrary offsets, ensuring that we don't zero out something that has been (or is in the middle of being) written. This scheme can also be applied to existing pools (affecting only free regions on the vdev). Detailed design:

- new subcommand:zpool initialize [-cs] <pool> [<vdev> ...]
    - start, suspend, or cancel initialization
- Creates new open-context thread for each vdev
- Thread iterates through all metaslabs in this vdev
- Each metaslab:
    - select a metaslab
    - load the metaslab
    - mark the metaslab as being zeroed
    - walk all free ranges within that metaslab and translate them to ranges on the leaf vdev
    - issue a "zeroing" I/O on the leaf vdev that corresponds to a free range on the metaslab we're working on
    - continue until all free ranges for this metaslab have been "zeroed"
    - reset/unmark the metaslab being zeroed
    - if more metaslabs exist, then repeat above tasks.
    - if no more metaslabs, then we're done.

- progress for the initialization is stored on-disk in the vdev’s leaf zap object. The following information is stored:
    - the last offset that has been initialized
    - the state of the initialization process (i.e. active, suspended, or canceled)
    - the start time for the initialization

- progress is reported via the zpool status command and shows information for each of the videos that are initializing

### How Has This Been Tested?

- Locally ran the ZTS, all tests pass.
- Locally ran `ztest` without failures
- Manually tested on a large-ish pool of 40 HDDs.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

dweeezil commented:
@behlendorf I'll get this reviewed over the pre-new-year weekend.

ikozhukhov commented:
how it is different with illumos version? i have tested illumos version on DilOS as well with some @grwilson helps :)

behlendorf commented:
@ikozhukhov the change is basically the same as applied to illumos.  @dweeezil made minimal changes to the test cases and a few other tweaks to make it Linux friendly.  

There are two followup commits 73526ba and 41158e8 which improve CLI performance when initializing more than a single vdev.  This was an important use case for us since we'd like to use this functionality to shake down new hardware for pools with 10's or 100's of vdevs.

Commit 2849493 was also added to address a cosmetic issue for those who manage their pools using using the CLI.

mailinglists35 commented:
>This scheme can also be applied to existing pools (affecting only free regions on the vdev).

does this mean the equivalent of manually running fstrim on ext4, finally?

kpande commented:
s/videos/vdevs/ in commit message

ahrens commented on cmd/zpool/zpool_main.c:
```diff
@@ -1923,6 +2041,49 @@ print_status_config(zpool_handle_t *zhp, status_cbdata_t *cb, const char *name,
 	}
 
+	/* Optionally display vdev initialization status for leaves */
+	if (cb->cb_print_vdev_init && children == 0) {
```
What specifically is the usability concern around displaying the init status by default?

When an initialization is in progress, how would admin's typically discover this?  I'm concerned that if they have to know to add the `-i` flag, they may miss this and not realize why their pool's performance is different during initialization.  We should consider printing the status by default if it is `VDEV_INITIALIZE_ACTIVE`.

behlendorf replied:
My specific concern is readability.  For anyone using a standard 80 column terminal the `pool status` output becomes needlessly challenging to read for the lifetime of the pool.

Though I can see the argument for wanting to see this while in the `VDEV_INITIALIZE_ACTIVE`state.  Why don't we print `(initializing)` when active, which is concise and consistent with the long standing behavior of printing `(resilvering)` or `(repairing)` after a vdev.  The new `-i` can then be used to report the detailed status.

ahrens replied:
That sounds good to me.

behlendorf commented:
@ahrens refreshed with your requested change for `zpool status`.

@kpande thanks for catching that, fixed.

@grwilson in the interests of making it easy to upstream and review the performance improvement for multiple vdevs I've squashed your changes and mine in to a single commit and added both of our signed-off-bys.  If you'd rather keep them separate just let me know and I can split them up again.

loli10K commented on include/libzfs_core.h:
```diff
@@ -62,6 +62,8 @@ int lzc_destroy_bookmarks(nvlist_t *, nvlist_t **);
 int lzc_load_key(const char *, boolean_t, uint8_t *, uint_t);
 int lzc_unload_key(const char *);
 int lzc_change_key(const char *, uint64_t, nvlist_t *, uint8_t *, uint_t);
+int lzc_initialize(const char *, pool_initialize_func_t, nvlist_t *,
```
should we also update pyzfs to support (and test) `lzc_initialize()`?

behlendorf replied:
Yes, yes we should.  Though, let's handle that in a follow up PR.

loli10K commented on module/zfs/spa.c:
```diff
@@ -6357,6 +6381,117 @@ spa_vdev_detach(spa_t *spa, uint64_t guid, uint64_t pguid, int replace_done)
+
+	for (nvpair_t *pair = nvlist_next_nvpair(nv, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(nv, pair)) {
+		uint64_t vdev_guid = fnvpair_value_uint64(pair);
```
This `fnvpair_value_uint64()` call can be used to panic the kernel from userland (though only root should be able to do it):

```
root@linux:/usr/src/zfs# cat /proc/sys/kernel/spl/gitrev
zfs-0.8.0-rc2-46-g9e4a98f
root@linux:/usr/src/zfs# git diff
diff --git a/lib/libzfs/libzfs_pool.c b/lib/libzfs/libzfs_pool.c
index f799471..a48cf0f 100644
--- a/lib/libzfs/libzfs_pool.c
+++ b/lib/libzfs/libzfs_pool.c
@@ -2146,7 +2146,7 @@ zpool_initialize(zpool_handle_t *zhp, pool_initialize_func_t cmd_type,
                }
 
                uint64_t guid = fnvlist_lookup_uint64(tgt, ZPOOL_CONFIG_GUID);
-               fnvlist_add_uint64(vdev_guids, vd_path, guid);
+               fnvlist_add_string(vdev_guids, vd_path, vd_path);
 
                (void) snprintf(msg, sizeof (msg), "%llu", (u_longlong_t)guid);
                fnvlist_add_string(guids_to_paths, msg, vd_path);
root@linux:/usr/src/zfs# zpool initialize testpool &
[1] 616
root@linux:/usr/src/zfs# 
Message from syslogd@linux at Jan  3 20:59:34 ...
 kernel:[   97.189554] VERIFY3(0 == nvpair_value_uint64(nvp, &rv)) failed (0 == 22)

Message from syslogd@linux at Jan  3 20:59:34 ...
 kernel:[   97.192212] PANIC at fnvpair.c:570:fnvpair_value_uint64()

root@linux:/usr/src/zfs# dmesg
[   67.876003] zavl: module license 'CDDL' taints kernel.
[   67.876008] Disabling lock debugging due to kernel taint
[   68.848431] [sched_delayed] sched: RT throttling activated
[   70.876414] ZFS: Loaded module v0.8.0-rc2_46_g9e4a98f (DEBUG mode), ZFS pool version 5000, ZFS filesystem version 5
[   83.079393] random: nonblocking pool is initialized
[   97.189554] VERIFY3(0 == nvpair_value_uint64(nvp, &rv)) failed (0 == 22)
[   97.192212] PANIC at fnvpair.c:570:fnvpair_value_uint64()
[   97.194322] Showing stack for process 616
[   97.194329] CPU: 0 PID: 616 Comm: zpool Tainted: P           O  3.16.0-4-amd64 #1 Debian 3.16.51-3
[   97.194333] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[   97.194335]  0000000000000000 ffffffff8151ea00 ffff880036e37700 ffff88003a707a80
[   97.194340]  ffffffffa0402421 ffff88003b6ab5d0 ffff88003a707c18 ffffffffa0402535
[   97.194343]  ffffffffa04a9e28 0000023a30bb2283 ffffffffa04aa6b0 ffffffffa04a8b18
[   97.194346] Call Trace:
[   97.194353]  [<ffffffff8151ea00>] ? dump_stack+0x5d/0x78
[   97.194363]  [<ffffffffa0402421>] ? spl_dumpstack+0x3c/0x3e [spl]
[   97.194369]  [<ffffffffa0402535>] ? spl_panic+0x112/0x14b [spl]
[   97.194375]  [<ffffffffa049f297>] ? nvt_add_nvpair+0xd5/0x1f4 [znvpair]
[   97.194382]  [<ffffffffa049f03a>] ? nvt_remove_nvpair+0x72/0x1fa [znvpair]
[   97.194389]  [<ffffffffa049eae2>] ? nvt_lookup_name_type+0x1be/0x2a8 [znvpair]
[   97.194398]  [<ffffffffa04a0ff1>] ? nvpair_value_common+0x2f/0x1b1 [znvpair]
[   97.194407]  [<ffffffffa04a228b>] ? nvpair_value_uint64+0x32/0x34 [znvpair]
[   97.194415]  [<ffffffffa04a65b0>] ? fnvpair_value_uint64+0x67/0x6d [znvpair]
[   97.194504]  [<ffffffffa067de92>] ? spa_vdev_initialize+0x1d7/0x329 [zfs]
[   97.194514]  [<ffffffffa04134a1>] ? spl_vmem_alloc+0x81/0x83 [spl]
[   97.194519]  [<ffffffffa04a66c2>] ? nv_alloc_sleep_spl+0x32/0x34 [znvpair]
[   97.194524]  [<ffffffffa049e444>] ? nv_mem_zalloc+0x5c/0x62 [znvpair]
[   97.194528]  [<ffffffffa049f53b>] ? nvlist_xalloc+0xa6/0xad [znvpair]
[   97.194533]  [<ffffffffa049f493>] ? nvlist_alloc+0x35/0x37 [znvpair]
[   97.194539]  [<ffffffffa04a4686>] ? fnvlist_alloc+0x2b/0x6c [znvpair]
[   97.194596]  [<ffffffffa071ed43>] ? zfs_ioc_pool_initialize+0x163/0x1c1 [zfs]
[   97.194645]  [<ffffffffa07268d3>] ? zfsdev_ioctl+0x4f5/0x799 [zfs]
[   97.194650]  [<ffffffff811c1b2f>] ? do_vfs_ioctl+0x2cf/0x4b0
[   97.194657]  [<ffffffff810593c7>] ? __do_page_fault+0x177/0x410
[   97.194663]  [<ffffffff811c1d91>] ? SyS_ioctl+0x81/0xa0
[   97.194668]  [<ffffffff81526d58>] ? async_page_fault+0x28/0x30
[   97.194675]  [<ffffffff81524d0d>] ? system_call_fast_compare_end+0x10/0x15
root@linux:/usr/src/zfs# 
```

behlendorf replied:
Good catch, I'll add a patch to address this.

loli10K commented on cmd/ztest/ztest.c:
```diff
@@ -374,6 +375,8 @@ ztest_func_t ztest_spa_upgrade;
 ztest_func_t ztest_remap_blocks;
 ztest_func_t ztest_spa_checkpoint_create_discard;
+ztest_func_t ztest_initialize;
+ztest_func_t ztest_initialize;
```
mismerge? this line is duplicate.

behlendorf replied:
Fixed.

loli10K commented on module/zfs/spa.c:
```diff
@@ -7025,6 +7184,22 @@ spa_async_thread(void *arg)
+		mutex_exit(&spa_namespace_lock);
+	}
+
+	if (tasks & SPA_ASYNC_INITIALIZE_RESTART) {
```
mismerge? this seems to be duplicate.

behlendorf replied:
Fixed

loli10K commented on cmd/zpool/zpool_main.c:
```diff
@@ -479,6 +505,97 @@ usage(boolean_t requested)
+		switch (c) {
+		case 'c':
+			if (cmd_type != POOL_INITIALIZE_DO) {
+				(void) fprintf(stderr, gettext("-c cannot be "
```
nit: unlike other zpool commands this does not seem to handle duplicate options

```
root@linux:~# zpool initialize testpool
root@linux:~# zpool initialize -s -s testpool
-s cannot be combined with other options
usage:
	initialize [-cs] <pool> [<device> ...]
root@linux:~# zpool initialize -c -c testpool
-c cannot be combined with other options
usage:
	initialize [-cs] <pool> [<device> ...]
root@linux:~# zpool scrub -s -s testpool
cannot cancel scrubbing testpool: there is no active scrub
root@linux:~# zpool scrub -p -p testpool
cannot pause scrubbing testpool: there is no active scrub
```

Also the synopsis seems to suggest both `-c` and `-s` can be used simultaneously, maybe something like `[ -c | -s ]` instead of `[-cs]` would be better.

behlendorf replied:
Fixed

behlendorf commented on module/zfs/vdev_queue.c:
```diff
@@ -154,6 +154,8 @@ uint32_t zfs_vdev_scrub_min_active = 1;
 uint32_t zfs_vdev_scrub_max_active = 2;
 uint32_t zfs_vdev_removal_min_active = 1;
 uint32_t zfs_vdev_removal_max_active = 2;
+uint32_t zfs_vdev_initializing_min_active = 1;
```
Added, I also added `zfs_vdev_removal_min_active` and `zfs_vdev_removal_max_active` which were missing.

behlendorf commented:
Rebased and updated to address the remaining nits.  I can squash those additional fixes in the final merge but for the moment kept them separate for easier review.
